SPI API: Adding  3-wire mode support

tbursztyka opened the pull request:
I had 2 options:

- keep the same size of spi_config structure and take a bit from the slave attribute
- rework the spi_config structure 

I took the 2nd option because C bit-fields is a no-go on a public API (if I had to go with C bit-fields) or parsing operation on one attribute and some in another would have been really ugly.
Also because the lines mode (single/dual/quad/octal) are not really used anywhere: such modes require some additional changes on the SPI API to be properly used (xec SPI driver tried to use it as it is but it's not exactly right, as writing a command forces the user to configure the controller on single line, holding CS, then re-configure on dual/quad/octal in order to get the command result at the right mode: it's heavy and cumbersome and limit the possibilities as well of these modes).
To really enable these modes something like #20564 should be done.

This obviously breaks struct spi_config a bit, but macros have same names, full duplex is by default (since it's 99.999% of the cases anyway) etc...

Commit:
spi: Add a configuration option for half/full duplex mode

This is meant to expose half/full duplex configuration option on regular
configuration. So far, dual/quad/octal are not exactly supported, as it
would require extensions to the SPI buffer for a full support.

So moving these modes to an extended operation attribute
(32 vs 16 bits), disabled by default.

And exposing half/full duplex configuration bit. Full duplex being the
default option.

Fixes #19134

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers: Apply SPI API change to relevant places

Mostly drivers, but also one sample and one test.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/sensors: Enable 3-wire SPI access to HTS221 sensor driver

HTS221 is a humidity and temperature sensor (thus HTS) that can be wired
on i2c or SPI bus. On SPI bus however, it uses the 3-wire mode, aka:
half-duplex.

Now that SPI API exposes half duplex operation, let's enable the SPI bus
on that sensor.

Let's move to a better DTS integrated driver as well, and also use
stmemsc interface.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
dts/bindings: Update HTS221 bindings to enable both i2c and SPI bus

HTS221 is now supported on both buses so let's make DTS aware of it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

avisconti commented:
@tbursztyka                                                                                 
                                                                                
It seems a very good job! Before looking at it in a more detailed way I have a question: did you use hts221 sensor to test SPI 3 wire? Have you also tested other sensors?

tbursztyka commented:
@avisconti  I am in the process of testing hts221 for this purpose, this is the only 3wire sensor I have at hand. But if it works for it, it should work for the others as well (knowing that other SPI controllers than STM32 will need to be updated as well).

tbursztyka commented:
Looks like some sensors can do full or half duplex, such as lis2dh (another ST sensor). I guess this will need a DTS solution in order to differentiate one or the other. (currently the spi version of this driver is obviously hard-coding full duplex mode).

avisconti commented:
> @avisconti I am in the process of testing hts221 for this purpose, this is the only 3wire sensor I have at hand. But if it works for it, it should work for the others as well (knowing that other SPI controllers than STM32 will need to be updated as well).

I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you,  if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?                                                                                   
Those routines are common among many ST sensors (mostly SPI 4-wires and I2C) and would be nice to have that part tested as well. If I'm requesting too much, I can do it later.                                                                               
                                                                               
And you may also make use of the latest DT helpers like I did in #38343 (not merged yet), in particular in commit c74640bceee66fbfa87bc019b23936225d4da2bf.  Please let me know.

avisconti commented:
> Looks like some sensors can do full or half duplex, such as lis2dh (another ST sensor). I guess this will need a DTS solution in order to differentiate one or the other. (currently the spi version of this driver is obviously hard-coding full duplex mode).

Yes, I agree.

tbursztyka commented:
> I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you, if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?

Sound relevant to do it now on hts221

avisconti commented:
> > I saw that you are also implementing multi-instance, which is good! In such a casa may I ask you, if not bothering too much, to use the common stmemsc API you can find under sensor/stmemsc ?
> 
> Sound relevant to do it now on hts221

@tbursztyka  Yes, and I guess you should cerry-pick c74640bceee66fbfa87bc019b23936225d4da2bf because the PR is still not merged.

scottwcpg commented:
@tbursztyka I checked with Microchip HW team and the MCHP XEC QMSPI controller can only support the 3 pin half-duplex mode if the QMSPI MOSI and MISO pins are externally connected together. Need to add driver logic to return an error if the driver is configured for 3-pin mode.

tbursztyka commented:
> @tbursztyka I checked with Microchip HW team and the MCHP XEC QMSPI controller can only support the 3 pin half-duplex mode if the QMSPI MOSI and MISO pins are externally connected together. 

That's always the case for all SPI controller as far as I know, so that's totally fine.

> Need to add driver logic to return an error if the driver is configured for 3-pin mode.

At first yes, I'll update all drivers so they reject SPI_HALF_DUPLEX, but it would be nice that each SPI driver owner would enable the support relevantly afterwards.

scottwcpg commented:
@tbursztyka Yes, once the PR is merged I will implement support in the XEC SPI driver(s).  

Somewhat off topic: Is there any chance your PR for dual/quad SPI or another will be merged?  Or has the Zephyr community decided dual/quad should be handled by a the flash driver?  If I implement a flash driver using the same QMSPI controller I will have "duplicate" drivers because the SPI driver has no support for dual/quad.  You are correct the dual/quad support I put in the XEC SPI driver is an ugly hack.

mbolivar-nordic commented:
Isn't this a stable API change? It has to go through the stable API change process if so.

Commit:
drivers/spi: Return an error on SPI_HALF_DUPLEX for relevant drivers

This feature will need to be, however, implemented driver by driver
afterwards.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

mbolivar-nordic reviewed (dismissed):
> Isn't this a stable API change? It has to go through the stable API change process if so.

I would like an answer to this question.

tbursztyka commented:
> > Isn't this a stable API change? It has to go through the stable API change process if so.

Yes thus why I put the DNM label. I want first to see users reaction, then once it's tested and proven to work: I'll propose the change through https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#stable (And I'll probably push for an updated #20564 as well)

> 
> I would like an answer to this question.

time zone, anyone?  ;)

mbolivar-nordic commented:
> > > Isn't this a stable API change? It has to go through the stable API change process if so.
> 
> Yes thus why I put the DNM label.

Great, thanks; I'm dropping my -1 on that understanding then.

avisconti commented on drivers/sensor/hts221/hts221.c:
```diff
@@ -180,17 +196,94 @@ int hts221_init(const struct device *dev)
+			.bus = DEVICE_DT_GET(DT_INST_BUS(inst)),	\
+			.spi_cfg = SPI_CONFIG_DT_INST(inst,		\
+					   HTS221_SPI_OPERATION,	\
+					   0),				\
```
Here is better to use SPI_DT_SPEC_INST_GET() and, similarly, I2C_DT_SPEC_INST_GET() for i2c bus. I'm saying so because stmemsc is going to change i/f because of #38343, which is not merged yet, so, depending which one will be merged first the other should be aligned. Not sure what the best option is. If this one will be merged first I will then change my PR, otherwise you can align this one from now (and cherry-pick my commit c74640bceee66fbfa87bc019b23936225d4da2bf, which can be merged with this PR instead of mine). What do you think?

avisconti replied:
Maybe @gmarull and @mbolivar can provide their feedback about how to proceed.

tbursztyka replied:
You PR will probably land for the next release, same for this PR as it has to go through API change process. So there will be time for adapting this PR against the changes on stmemsc.

tbursztyka replied:
@avisconti change was applied btw.

avisconti replied:
> @avisconti change was applied btw.

OK, I'm reviewing right now

mbolivar-nordic commented on include/drivers/spi.h:
```diff
@@ -246,8 +260,14 @@ struct spi_cs_control {
 struct spi_config {
 	uint32_t		frequency;
+#if defined(CONFIG_SPI_EXTENDED_MODES)
+	uint32_t		operation;
```
Is client code expected to ifdef the size of the operation value too? Maybe it's worth adding a `typedef` for this. That would avoid some duplication here too.

Just an idea.

tbursztyka replied:
it will not need to ifdef for the size but for using the extended modes configuration bits (if it supports both normal single line or multiple ones). At least I don't see the size as an issue.

Commit:
dts/bindings: Add a dedicated property for duplex mode on SPI device

Users will have to include dt-bindings/spi/spi.h in order to use the
relevant flags fol this property. For instance:

    ...
    duplex = <SPI_HALF_DUPLEX>;
    ....

By default all SPI device are configured to be full duplex so the
property is optional. This property makes sense only for devices that
can be configured on either modes. Which, in such case, it will need to
use DT_INST_PROP(<instance number>, duplex) macro call to retrieve the
property value. Others can fully ignore it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
@avisconti  can you re-review please?

@mbolivar-nordic Could you re-review, especially the last commit I added. There is no device at the moment that will use such property in DTS, but I hope this is the best way to expose it.

carlescufi commented:
@anangl could you take a look please?

avisconti reviewed (approved):
I just checked the hts221 sensor changes, and it seems a very good job! Thx!

mbolivar-nordic commented on dts/bindings/spi/spi-device.yaml:
```diff
@@ -16,3 +16,7 @@ properties:
       required: true
+    duplex:
+      type: int
+      required: false
+      description: Duplex mode, full or half. By default it's always full.
```
I think this should use an enum (https://docs.zephyrproject.org/latest/guides/dts/bindings.html#enum-values) and a default (0), with a justification for why (that's the usual case for SPI controllers).

mbolivar-nordic reviewed (changes requested):
One minor request for the binding.

anangl commented on drivers/spi/Kconfig:
```diff
@@ -26,6 +26,15 @@ config SPI_SLAVE
+	select EXPERIMENTAL
+	help
+	  Enables extended operations in the SPI API. Currently, this
+	  enables the possibility to select the line mode (singl/dual/
```
```suggestion
	  enables the possibility to select the line mode (single/dual/
```

anangl commented on include/drivers/spi.h:
```diff
@@ -228,10 +230,14 @@ struct spi_cs_control {
+ *     cs_active_high      [ 14 ]      - Active high CS logic.
+ *     reserved            [ 15 ]      - reserved for future use.
+ * if CONFIG_SPI_EXTENDED_MODES is defined:
+ *     lines               [ 16 : 18 ] - MISO lines: Single/Dual/Quad/Octal.
+ *     reserved            [ 19 : 31 ] - reserved for future use.
```
```suggestion
 *     lines               [ 16 : 17 ] - MISO lines: Single/Dual/Quad/Octal.
 *     reserved            [ 18 : 31 ] - reserved for future use.
```

tbursztyka replied:
indeed

gmarull commented on include/drivers/spi.h:
```diff
@@ -116,7 +112,21 @@ extern "C" {
+ * Default is single, which is the case most of the time.
+ * Without CONFIG_SPI_EXTENDED_MODES being enabled, single is the only supported one.
+ */
+#define SPI_LINES_SINGLE	(0 << 16)
+#define SPI_LINES_DUAL		(1 << 16)
+#define SPI_LINES_QUAD		(2 << 16)
+#define SPI_LINES_OCTAL		(3 << 16)
+
+#define SPI_LINES_MASK		(0x3 << 16)
```
nit: add `U` suffix to all unsigned numbers

gmarull commented on include/drivers/spi.h:
```diff
@@ -116,7 +112,21 @@ extern "C" {
+#define SPI_CS_ACTIVE_HIGH	BIT(14)
+
+/**
+ * @brief SPI MISO lines (if only CONFIG_SPI_EXTENDED_MODES is enabled)
```
nit: this will just document first define, not the others. You can use

```c
/**
 * @name SPI MISO lines ....
 * @{
 */

...

/** @} */
```

gmarull commented on include/drivers/spi.h:
```diff
@@ -228,10 +238,14 @@ struct spi_cs_control {
+ *     lock_on             [ 13 ]      - Keep resource locked for the caller.
+ *     cs_active_high      [ 14 ]      - Active high CS logic.
+ *     reserved            [ 15 ]      - reserved for future use.
+ * if CONFIG_SPI_EXTENDED_MODES is defined:
```
`@kconfig{CONFIG_SPI_EXTENDED_MODES}`

gmarull commented on drivers/spi/spi_litespi.c:
```diff
@@ -40,7 +40,8 @@ static int spi_config(const struct spi_config *config, uint16_t *control)
 
-	if ((config->operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
+	if (IS_ENABLED(CONFIG_SPI_EXTENDED_MODES) &&
+	    (config->operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
```
nit: would be nice to provide some helpers, e.g.

```c
#define SPI_GET_LINES(operation) ((operation) & SPI_LINES_MASK)
```

to then do:

```c
SPI_GET_LINES(config->operation) != SPI_LINES_SINGLE) {
```

tbursztyka replied:
Let's no that in #39991

gmarull commented on drivers/sensor/hts221/hts221.c:
```diff
@@ -10,16 +10,22 @@
+static const struct str2odr hts221_odrs[] = {
+	{ "1", HTS221_ODR_1Hz },
+	{ "7", HTS221_ODR_7Hz },
+	{ "12.5 ", HTS221_ODR_12Hz5 },
```
nit: extra space after `12.5`

gmarull commented on drivers/sensor/hts221/hts221_trigger.c:
```diff
@@ -121,33 +118,46 @@ int hts221_init_interrupt(const struct device *dev)
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int status;
+
+	/* setup data ready gpio interrupt */
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
+		/* API may return false even when ptr is NULL */
+		if (cfg->gpio_drdy.port != NULL) {
+			LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
+			return -ENODEV;
+		}
+
+		LOG_DBG("gpio_drdy not defined in DT");
+		return 0;
+	}
```
may read better if:

```c 
	if (cfg->gpio_drdy.port == NULL) {
		LOG_DBG("gpio_drdy not defined in DT");
		return 0;
	}

	if (!device_is_ready(cfg->gpio_drdy.port)) {
		LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
		return -ENODEV;
	}
```

tbursztyka replied:
I just c/p what has been done in other drivers but indeed it's more logical to read it that way

gmarull commented on dts/bindings/spi/spi-device.yaml:
```diff
@@ -16,3 +16,15 @@ properties:
+    duplex:
+      type: int
+      required: false
+      description: Duplex mode, full or half. By default it's always full
```
nit for multiline description: use `description: |`

mbolivar-nordic replied:
Seconded

tbursztyka replied:
only if you break the line right after description:  or for all? 

I saw confusing examples in other bindings. Anyway I'll break the line.

mbolivar-nordic replied:
> only if you break the line right after description: or for all?

The rules and rationale for them are documented in 
https://docs.zephyrproject.org/latest/guides/dts/bindings.html#general-rules

gmarull commented on include/dt-bindings/spi/spi.h:
```diff
@@ -0,0 +1,29 @@
+ * @{
+ */
+
+/**
+ * @brief SPI duplex mode
+ *
+ * Some controllers support half duplex transfer, which results in 3-wire usage.
+ * By default, full duplex will prevail.
+ */
+#define SPI_FULL_DUPLEX		(0 << 11)
+#define SPI_HALF_DUPLEX		(1 << 11)
```
only first define gets documented, see related previous code

tbursztyka replied:
Actually this is on all multiple defines, I'll make a dedicated commit for it

henrikbrixandersen reviewed (changes requested):
Looks good, thank you. I don't have any other findings except for the ones already reported by @gmarull. I will approve once those have been fixed.

Commit:
spi: Precise suffix to all unsigned numbers

The rule was not applied on spi header.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
spi: Fixing documentation for groups of macros

Group macros documentation properly.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

carlescufi commented:
@mbolivar-nordic please revisit

carlescufi commented:
@henrikbrixandersen please revisit

henrikbrixandersen commented on drivers/sensor/hts221/hts221_trigger.c:
```diff
@@ -121,33 +118,44 @@ int hts221_init_interrupt(const struct device *dev)
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	int status;
+
+	if (cfg->gpio_drdy.port != NULL) {
+		LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(cfg->gpio_drdy.port)) {
+		LOG_DBG("gpio_drdy not defined in DT");
+		return 0;
+	}
```
```suggestion
	if (cfg->gpio_drdy.port == NULL) {
		LOG_DBG("gpio_drdy not defined in DT");
		return 0;
	}

	if (!device_is_ready(cfg->gpio_drdy.port)) {
		LOG_ERR("device %s is not ready", cfg->gpio_drdy.port->name);
		return -ENODEV;
	}
```

kapi-no commented on drivers/sensor/hts221/hts221.h:
```diff
@@ -10,29 +10,23 @@
 
-#define HTS221_TRIGGER_ENABLED (DT_INST_NODE_HAS_PROP(0, drdy_gpios) && \
-				IS_ENABLED(CONFIG_HTS221_TRIGGER))
+#include "hts221_reg.h"
```
@tbursztyka, it looks like you forgot to commit this header file
